src/lib/mina_lib/mina_lib.ml

```diff
@@ -1493,6 +1493,95 @@ let start_filtered_log ~commit_id
       () ;
     Ok () )
 
+let fetch_completed_snarks (module Context : CONTEXT) snark_pool network =
+  let open Context in
+  let open Network_peer in
+  let%bind all_peers = Mina_networking.peers network in
+  Deferred.List.iter
+    ~f:(fun peer ->
+      [%log debug] "PEER IS: Fetching completed snarks from peer: $peer"
+        ~metadata:[ ("peer", Network_peer.Peer.to_yojson peer) ] ;
+      let completed_works =
+        Mina_networking.get_completed_checked_snarks network peer
+      in
+      let%bind completed_works = completed_works in
+      let completed_works =
+        match completed_works with
+        | Error e ->
+            [%log debug]
+              ~metadata:
+                [ ("peer", Network_peer.Peer.to_yojson peer)
+                ; ("error", Error_json.error_to_yojson e)
+                ]
+              "Failed to fetch completed snarks from peer: $error" ;
+            []
+        | Ok completed_works ->
+            completed_works
+      in
+      let%bind () =
+        Deferred.List.iter completed_works ~f:(fun work ->
+            let callback =
+              let cb =
+                Mina_net2.Validation_callback.create_without_expiration ()
+              in
+              Network_pool.Snark_pool.Broadcast_callback.External cb
+            in
+
+            (* proofs should be verified in apply and broadcast *)
+            let statement = Transaction_snark_work.Checked.statement work in
+            let snark =
+              Network_pool.Priced_proof.
+                { proof = work.proofs
+                ; fee = { fee = work.fee; prover = work.prover }
+                }
+            in
+            let msg =
+              let diff =
+                Network_pool.Snark_pool.Diff_versioned.Add_solved_work
+                  (statement, snark)
+              in
+              Envelope.Incoming.wrap_peer ~data:diff ~sender:peer
+            in
+            (* verify the snarks to be added *)
+            let resource_pool =
+              Network_pool.Snark_pool.resource_pool snark_pool
+            in
+            let%bind err =
+              Network_pool.Snark_pool.Resource_pool.verify_and_act resource_pool
+                ~work:(statement, snark) ~sender:msg.sender
+            in
+            match err with
+            | Ok () ->
+                [%log debug]
+                  ~metadata:
+                    [ ("peer", Network_peer.Peer.to_yojson peer)
+                    ; ( "work_ids"
+                      , Transaction_snark_work.Statement.compact_json statement
+                      )
+                    ]
+                  "Successfully verified snark work from peer: $peer" ;
+
+                (* does an empty check for the snark, then an unsafe apply, and finally adds it to the pool *)
+                Network_pool.Snark_pool.apply_and_broadcast snark_pool msg
```

Comment:
Broadcasting here doesn't make sense to me. Nodes periodically syncing by requesting work might work better if the goal is to sync snark pool on a regular basis.